### PR TITLE
Created Several classes for User Manager Implementation.

### PR DESCRIPTION
**Pull request by JayBSB**
Created UserManager class to handle logging in, registering and handling of users. Created Permission_Object class due to enums being insufficient for DI. Created User_Tester class to ensure implementation was functional. IUser and Permission.cs are possibly redundant now but wanted to discuss with team further implementation before deletion.

UserManager can either be refactored to work another way, to tie in with StateManager to authenticate and access User accounts. Created an abstract user class to encapsulate functionality with UserManager, but will refactor if anyone can think of an easier way. Sorry I had to change so much. ):

### CODE REVIEW

File: PMSCore/Permission_Object.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        /// </summary>
+        /// <returns>A string representation of the user's permission level.</returns>
+        public string GetPermissionLevel()
+        {
+            // Log the access of the permission level.
+            _logger.LogInformation("Permission Accessed: {ThisPermission}", _permissionLevel.ToString());
+            return _permissionLevel.ToString();
+        }
+    }
+}
```
Comment (DistroRickDev) [inline PMSCore/Permission_Object.cs:65]: Permission object should be an actual object in a XUser class, don't see the need for it

File: PMSCore/User.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
+﻿namespace PMSCore
+{
+    internal abstract class User
+    {
+        private Permission_Object _UserPermission; // Stores the permission object for the user.
```
Comment (DistroRickDev) [inline PMSCore/User.cs]: This should be of the type Permission the enum created with the IUser interface

File: PMSCore/User.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
+﻿namespace PMSCore
+{
+    internal abstract class User
```
Comment (DistroRickDev) [inline PMSCore/User.cs]: why do we have this internal abstract class? also shouldn't it inherit from IUser ?

File: PMSCore/User.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+                _logger.LogDebug($"{_userID}'s PERMISSION LEVEL: {_UserPermission.GetPermissionLevel()}");
+
+            }
+            catch (Exception ex)
+            {
+                //TODO Create custom exceptions for failures.
+
+                // Log critical error during user initialization.
+                _logger.LogCritical("{datetime}: Invalid User.\n{errorDetails}", DateTime.Now, ex);
+            }
```
Comment (DistroRickDev) [inline PMSCore/User.cs]: why a try catch in a class constructor? what is supposed to throw here?

File: PMSCore/User.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        }
+
+        /// <summary>
+        /// Retrieves the user's permission level.
+        /// </summary>
+        /// <returns>A string representing the user's permission level.</returns>
+        public string ViewUserPermissions()
+        {
+            return this._UserPermission.GetPermissionLevel();
```
Comment (DistroRickDev) [inline PMSCore/User.cs]: Permission.ToString() and this should return the type itself

File: PMSCore/User.cs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        }
+
+        /// <summary>
+        /// Retrieves the user's ID.
+        /// </summary>
+        /// <returns>The user ID as a string.</returns>
+        public string GetUserID()
+        {
+            return this._userID;
```
Comment (DistroRickDev) [inline PMSCore/User.cs]: this is redudant

File: PMSCore/User_Tester.cs [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+        /// Constructor for User_Tester. Passes logger and permission object to the base class.
+        /// </summary>
+        /// <param name="logger">Optional logger instance.</param>
+        /// <param name="permission">Permission object for the user.</param>
+        internal User_Tester(ILogger? logger, Permission_Object permission, string userID) : base(logger, permission, userID)
+        {
+            _logger = logger ?? new LoggerFactory().CreateLogger("User_Tester");
+
+        }
+    }
```
Comment (DistroRickDev) [inline PMSCore/User_Tester.cs]: is this suppose to be a mock or fake ? It already exsits one UserFake in the codebase for testing

File: PMSCore/Utils/UserManager.cs [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+using PMSCore;
+
+namespace PMSCore.Utils
+{
+    internal class UserManager
+
+    {
+        private User _currentUser; // Stores the currently logged-in user.
+        private readonly ILogger _logger; // Logger for the UserManager class.
+
```
Comment (DistroRickDev) [inline PMSCore/Utils/UserManager.cs]: maybe we missing a list of users as well so we can list all Users in the system

File: PMSCore/Utils/UserManager.cs [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+        {
+            // Use the provided logger or create a new one if null.
+            _logger = logger ?? new LoggerFactory().CreateLogger("User_Tester");
+            _logger.LogDebug("Setting up User Manager.");
+
+            // Attempt login; if it fails, register a new user.
+            if (!this.Login())
+            {
+                this.RegisterNewUser();
+            }
```
Comment (DistroRickDev) [inline PMSCore/Utils/UserManager.cs]: this should be in it's own method

File: PMSCore/Utils/UserManager.cs [hunk 10]
```diff
@@ -0,0 +1,138 @@
+﻿using Microsoft.Extensions.Logging;
+using PMSCore;
+
+namespace PMSCore.Utils
+{
+    internal class UserManager
```
Comment (DistroRickDev) [inline PMSCore/Utils/UserManager.cs]: this class should be a singletone we only need a user manager instance in the entire project scope

File: PMSCore/Utils/UserManager.cs [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+
+        }
+
+        /// <summary>
+        /// Retrieves the permission level of the current user.
+        /// </summary>
+        /// <returns>A string representing the user's permissions.</returns>
+        public string GetPermission()
+        {
+            return _currentUser.ViewUserPermissions();
```
Comment (DistroRickDev) [inline PMSCore/Utils/UserManager.cs]: GetPermissionLevel()

File: PMSCore/Utils/UserManager.cs [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+                    (
+                    new LoggerFactory()
+                    .CreateLogger("User_Tester"),
+                    new Permission_Object(2),
+                    "J-Doe"
+                    );
+            }
+
+            return _currentUser != null; // Return true if user is successfully created.
+        }
```
Comment (DistroRickDev) [inline PMSCore/Utils/UserManager.cs]: Login should receive a user object, if it exists in the list current_user becomes that user and returns a state else current_user remains what is was and an error status gets returned

File: PMSCore/Utils/UserManager.cs [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+            _currentUser = new User_Tester
+                (
+                new LoggerFactory()
+                .CreateLogger("User_Tester"),
+                new Permission_Object(2),
+                "J-Doe"
+                );
+
+            return _currentUser != null; // Return true if user is successfully created.
+        }
```
Comment (DistroRickDev) [inline PMSCore/Utils/UserManager.cs]: Register should receive a user object, if it exists in the list current_user returns an error else adds it to the list

**Review by DistroRickDev — COMMENTED** [review]
I have a few questions that arise in this PR:
The focus should be in having a single entity that manages users, I dont see a need to create a user class here when the IUser interface should be used in order to achieve abstraction.
It seems u have created a test class in the src directory when there is a fake in the test directory for it in existance already
I believe this PR could be reworked to just PMSCore\UserManager.cs as a singleton (similar to https://github.com/DistroRickDev/PMS-Core/blob/master/PMSCore/ProjectManager.cs) and PMSCore.Test\UserManagerTest.cs

**Comment by JayBSB** [comment]
Did that thing again when I accidentally-on-purpose (did it but forgot I wasn't supposed to) merged main into my branch again, which meant I couldn't squash the commits. Reapplied changes on a new branch, will open pull request for that now instead.

**Pull request closed without merging**